### PR TITLE
feat(auth0): migrate auth to auth0

### LINKED ISSUE #17114: Update the signup to include a checkbox for agreeing to ToS and Privacy and a checkbox for receiving Quincy's emails.

### LINKED ISSUE #17147: Replace Loopback auth with Auth0 SSO
At the minute, loopback assigns an access token to each user for their current session. This is only good for one domain.

We need to replace this auth process with Auth0 Single Sign On. This will allow tokens to be carried between different subdomains like learn, guide and the forum.

Something to be cleared up is, if a user signs out on the `www` domain, how is this reflected on other subdomains that still hold valid token in `localStorage`?

### PR DESCRIPTION
**Pull request by raisedadead**
Closes #17147
Closes #17114

### CODE REVIEW

**Comment by raisedadead** [comment]
Hmm so the Social auths are broken.. but that's fine. I will fix up in the morning.

File: common/models/User-Identity.js [hunk 1]
```diff
@@ -15,6 +14,7 @@ import { wrapHandledError } from '../../server/utils/create-handled-error.js';
 export default function(UserIdent) {
   UserIdent.on('dataSourceAttached', () => {
     UserIdent.findOne$ = observeMethod(UserIdent, 'findOne');
+    UserIdent.create$ = observeMethod(UserIdent, 'create');
```
Comment (raisedadead) [inline common/models/User-Identity.js]: This line is no longer in use. Remove it.

File: common/models/User-Identity.js [hunk 2]
```diff
@@ -48,84 +48,98 @@ export default function(UserIdent) {
...
+              {
+                isOrphaned: username || true
+              }
+            )
+              .do(() => {
+                throw wrapHandledError(
+                  new Error('user identity is not associated with a user'),
+                  {
+                    type: 'info',
+                    redirectTo: '/deprecated-signup',
```
Comment (raisedadead) [inline common/models/User-Identity.js]: `/deprecated-signup` => `/signup`

File: common/models/User-Identity.js [hunk 3]
```diff
@@ -48,84 +48,98 @@ export default function(UserIdent) {
...
+          if (!identity) {
+            throw wrapHandledError(
+              new Error('user identity account not found'),
+              {
+                message: dedent`
+                  New accounts can only be created using an email address.
+                  Please create an account below
+                `,
+                type: 'info',
+                redirectTo: '/deprecated-signup'
```
Comment (raisedadead) [inline common/models/User-Identity.js]: `/deprecated-signup` => `/signup`

File: server/component-passport.js [hunk 4]
```diff
@@ -90,8 +91,58 @@ export default function setupPassport(app) {
...
+    let successRedirect = (req) => {
+      if (!!req && req.session && req.session.returnTo) {
+        let returnTo = req.session.returnTo;
+        delete req.session.returnTo;
+        return returnTo;
+      }
+      return config.successRedirect || '';
+    };
+
+    config.customCallback = !config.redirectWithToken
```
Comment (raisedadead) [inline server/component-passport.js]: This flag is misleading

File: server/passport-providers.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 const successRedirect = '/';
-const failureRedirect = '/signin';
+const failureRedirect = '/';
```
Comment (raisedadead) [inline server/passport-providers.js]: Revert this?

File: server/passport-providers.js [hunk 6]
```diff
@@ -165,5 +165,21 @@ export default {
...
+    provider: 'auth0',
+    module: 'passport-auth0',
+    clientID: process.env.AUTH0_CLIENT_ID,
+    clientSecret: process.env.AUTH0_CLIENT_SECRET,
+    domain: process.env.AUTH0_DOMAIN,
+    cookieDomain: 'freeCodeCamp.org',
+    callbackURL: '/auth/auth0/callback',
+    authPath: '/auth/auth0',
+    callbackPath: '/auth/auth0/callback',
+    redirectWithToken: true,
```
Comment (raisedadead) [inline server/passport-providers.js]: This flag is misleading

File: hosted-templates/hosted-pages/login.html [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+        var email = $('input#email').val();
+        var code = $('input#code').val();
+        webAuth.passwordlessLogin({
+          connection: 'email',
+          email: email,
+          verificationCode: code
+        }, function(err, res) {
+          if (err) {
+            console.log('error logging-in', err);
+            alert('Couldn\'t login ' + err.message);
```
Comment (raisedadead) [inline hosted-templates/hosted-pages/login.html]: Handle error properly.

File: hosted-templates/hosted-pages/login.html [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+          }
+          console.log('res', res);
+          // Hide the input and show the code entry screen
+          $('.enter-email').hide();
+          $('.enter-code').show();
+        });
+      }
+
+      function login() {
+        console.log('logging in...');
```
Comment (raisedadead) [inline hosted-templates/hosted-pages/login.html]: Remove log

File: hosted-templates/hosted-pages/login.html [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+        var email = $('input#email').val();
+        var code = $('input#code').val();
+        webAuth.passwordlessLogin({
+          connection: 'email',
+          email: email,
+          verificationCode: code
+        }, function(err, res) {
+          if (err) {
+            console.log('error logging-in', err);
```
Comment (raisedadead) [inline hosted-templates/hosted-pages/login.html]: Remove log

File: hosted-templates/hosted-pages/login.html [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+      function sendEmail() {
+
+        var email = $('input#email').val();
+        webAuth.passwordlessStart({
+          connection: 'email',
+          send: 'code',
+          email: email
+        }, function(err, res) {
+          if (err) {
+            // Handle error
```
Comment (raisedadead) [inline hosted-templates/hosted-pages/login.html]: Handle error properly. Remove log

**Review by Bouncey — APPROVED** [review]

**Pull request merged**